src/backend/graphics/opengl.rs

```diff
@@ -0,0 +1,87 @@
+//! Common traits and types for opengl rendering on graphics backends
+
+use nix::c_void;
+
+/// Error that can happen when swapping buffers.
+#[derive(Debug, Clone)]
+pub enum SwapBuffersError {
+    /// The OpenGL context has been lost and needs to be recreated.
+    ///
+    /// All the objects associated to it (textures, buffers, programs, etc.)
+    /// need to be recreated from scratch.
+    ///
+    /// Operations will have no effect. Functions that read textures, buffers, etc.
+    /// from OpenGL will return uninitialized data instead.
+    ///
+    /// A context loss usually happens on mobile devices when the user puts the
+    /// application on sleep and wakes it up later. However any OpenGL implementation
+    /// can theoretically lose the context at any time.
+    ContextLost,
+    /// The buffers have already been swapped.
+    ///
+    /// This error can be returned when `swap_buffers` has been called multiple times
+    AlreadySwapped,
+}
+
+/// All APIs related to OpenGL that you can possibly get
+/// through OpenglRenderer implementations
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Api {
+    /// The classical OpenGL. Available on Windows, Linux, OS/X.
+    OpenGl,
+    /// OpenGL embedded system. Available on Linux, Android.
+    OpenGlEs,
+    /// OpenGL for the web. Very similar to OpenGL ES.
+    WebGl,
+}
+
+/// Describes the pixel format of the main framebuffer
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PixelFormat {
+    /// Is the format hardware accelerated
+    pub hardware_accelerated: bool,
+    /// bits used for colors
+    pub color_bits: u8,
+    /// bits used for alpha channel
+    pub alpha_bits: u8,
+    /// bits used for depth channel
+    pub depth_bits: u8,
+    /// bits used for stencil buffer
+    pub stencil_bits: u8,
+    /// is stereoscopy enabled
+    pub stereoscopy: bool,
+    /// is double buffering enabled
+    pub double_buffer: bool,
+    /// multisampling format used if enabled
+    pub multisampling: Option<u16>,
+    /// if the format has srgb enabled
+    pub srgb: bool,
+}
+
+/// Trait that describes objects that have an OpenGl context
+/// and can be used to render upon
+pub trait OpenglRenderer
+{
+    /// Swaps buffers at the end of a frame.
+    fn swap_buffers(&self) -> Result<(), SwapBuffersError>;
+
+    /// Returns the address of an OpenGL function.
+    ///
+    /// Supposes that the context has been made current before this function is called.
+    unsafe fn get_proc_address(&self, symbol: &str) -> *const c_void;
+
+    /// Returns the dimensions of the window, or screen, etc.
```

Comment:
Others won't scale because there is literally no way to get a scaling screen with something like a drm backend. One points always equals one pixel in that case. For `glutin` there is actually a way to work around this. OpenGL contexts are some of the only APIs where you can get pixel but also point accuracy. So when you say, that we should scale for nested compositors like when using the `glutin` backend, but not for other, "screen points" is actually the correct term to refer to in the docs, because "pixels" would mean we don't scale on `glutin`.

Scaling the clients correctly is actually always something the compositor needs to handle.